src/Interpreters/Squashing.cpp

```diff
@@ -0,0 +1,159 @@
+#include <vector>
+#include <Interpreters/Squashing.h>
+#include <Common/CurrentThread.h>
+
+
+namespace DB
+{
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+Squashing::Squashing(Block header_, size_t min_block_size_rows_, size_t min_block_size_bytes_)
+    : header(header_)
+    , min_block_size_rows(min_block_size_rows_)
+    , min_block_size_bytes(min_block_size_bytes_)
+{
+}
+
+Chunk Squashing::flush()
+{
+    return convertToChunk(std::move(chunks_to_merge_vec));
+}
+
+Chunk Squashing::squash(Chunk && input_chunk)
+{
+    if (!input_chunk.hasChunkInfo())
+        return Chunk();
+
+    const auto *info = getInfoFromChunk(input_chunk);
+    return squash(info->chunks);
+}
+
+Chunk Squashing::add(Chunk && input_chunk)
+{
+    if (!input_chunk)
+        return {};
+
+    /// Just read block is already enough.
+    if (isEnoughSize(input_chunk.getNumRows(), input_chunk.bytes()))
+    {
+        /// If no accumulated data, return just read block.
+        if (chunks_to_merge_vec.empty())
+        {
+            chunks_to_merge_vec.push_back(std::move(input_chunk));
+            Chunk res_chunk = convertToChunk(std::move(chunks_to_merge_vec));
+            chunks_to_merge_vec.clear();
+            return res_chunk;
+        }
+
+        /// Return accumulated data (maybe it has small size) and place new block to accumulated data.
+        Chunk res_chunk = convertToChunk(std::move(chunks_to_merge_vec));
+        chunks_to_merge_vec.clear();
+        changeCurrentSize(input_chunk.getNumRows(), input_chunk.bytes());
+        chunks_to_merge_vec.push_back(std::move(input_chunk));
+        return res_chunk;
+    }
+
+    /// Accumulated block is already enough.
+    if (isEnoughSize(accumulated_size.rows, accumulated_size.bytes))
+    {
+        /// Return accumulated data and place new block to accumulated data.
+        Chunk res_chunk = convertToChunk(std::move(chunks_to_merge_vec));
+        chunks_to_merge_vec.clear();
+        changeCurrentSize(input_chunk.getNumRows(), input_chunk.bytes());
+        chunks_to_merge_vec.push_back(std::move(input_chunk));
+        return res_chunk;
+    }
+
+    /// Pushing data into accumulating vector
+    expandCurrentSize(input_chunk.getNumRows(), input_chunk.bytes());
+    chunks_to_merge_vec.push_back(std::move(input_chunk));
+
+    /// If accumulated data is big enough, we send it
+    if (isEnoughSize(accumulated_size.rows, accumulated_size.bytes))
+    {
+        Chunk res_chunk = convertToChunk(std::move(chunks_to_merge_vec));
+        changeCurrentSize(0, 0);
+        chunks_to_merge_vec.clear();
+        return res_chunk;
+    }
+    return {};
+}
+
+Chunk Squashing::convertToChunk(std::vector<Chunk> && chunks) const
+{
+    if (chunks.empty())
+        return {};
+
+    auto info = std::make_shared<ChunksToSquash>();
+    info->chunks = std::move(chunks);
+
+    chunks.clear();
+
+    return Chunk(header.cloneEmptyColumns(), 0, info);
```

Comment:
`header` is being used only here. If we create here just empty chunk without columns, nothing changes. But we cut dependencies.